metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/DimensionMapper.java

```diff
@@ -126,33 +131,46 @@ private DimensionMapper(
      * the dimension will be added with a value of "null".
      *
      * @param metric Metric being sent
-     * @param context TypedMap containing metric context attributes
+     * @param context the context in which the metric was recorded.
      * @return A list of Dimensions appropriate to send to CloudWatch
      */
-    public List<Dimension> getDimensions(final Metric metric, final TypedMap context)
+    public List<Dimension> getDimensions(final Metric metric, final MetricRecorder.RecorderContext context)
     {
+        TypedMap attributes = flattenContextHierarchy(context);
         Set<TypedMap.Key> dimensionKeys = filterMap.get(metric);
         if (dimensionKeys == null) {
             dimensionKeys = Collections.emptySet();
         }
 
-        Set<Dimension> dimensions = new HashSet<>(globalDimensions.size() + dimensionKeys.size());
+        Stream<TypedMap.Key> keys = Stream.concat(globalDimensions.stream(), dimensionKeys.stream());
+        return keys.map(key -> new Dimension().withName(key.name).withValue(String.valueOf(attributes.get(key))))
+            .distinct()
+            .sorted(Comparator.comparing(Dimension::getName))
+            .collect(Collectors.toList());
+    }
 
-        for (TypedMap.Key k : globalDimensions) {
-            Dimension d = new Dimension();
-            d.setName(k.name);
-            d.setValue(String.valueOf(context.get(k)));
-            dimensions.add(d);
+     /**
+     * Flatten the attributes in the context hierarchy to produce the full set of
+     * attributes which describe the environment in which measurements were taken.
+     * Parent attributes are overridden by child attributes with the same key.
+     *
+     * @param context the context used for recording measurements.
+     * @return the flattened attributes with parent values overriden by child.
+     */
+    private TypedMap flattenContextHierarchy(MetricRecorder.RecorderContext context) {
+        MetricRecorder.RecorderContext parent = context;
+        ImmutableTypedMap.Builder attributes = ImmutableTypedMap.Builder.from(TypedMap.empty());
+        Stack<MetricRecorder.RecorderContext> parents =  new Stack<>();
+        while(parent != null) {
+            parents.push(parent);
+            parent = parent.parent();
         }
 
-        for (TypedMap.Key k : dimensionKeys) {
-            Dimension d = new Dimension();
-            d.setName(k.name);
-            d.setValue(String.valueOf(context.get(k)));
-            dimensions.add(d);
+        while(!parents.empty()) {
+            MetricRecorder.RecorderContext current = parents.pop();
+            current.attributes().forEach(entry -> attributes.add(entry.getKey(), entry.getValue()));
```

Comment:
A comment here clarifying that this is where the attributes get over written would clarify the code a bit. I'd rather not rely on a comment at the function level when the behavior results in potential data loss.